[lldb][FreeBSD] Add dynamic loader handle class for FreeBSD Kernel

aokblast opened the pull request:
This commit is moved from [llvm-phabricator](https://reviews.llvm.org/D159076).
The implemtation support parsing kernel module for FreeBSD Kernel and has been test on x86-64 and arm64.
In summary, this class parse the linked list resides in the kernel memory that record all kernel module and load the debug symbol file to facilitate debug process

clayborg commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+// Instantiate Function of the FreeBSD Kernel Dynamic Loader Plugin called when
+// Register the Plugin
+DynamicLoader *
+DynamicLoaderFreeBSDKernel::CreateInstance(lldb_private::Process *process,
+                                           bool force) {
```
This function is important that we make sure when "force == false" tha we don't say that this plug-in can handle another Free BSD debugging session. I would feel better if "is_kernel()" was rock solid. See comments in that function for details

bulbazord commented on lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp:
```diff
@@ -935,6 +935,16 @@ lldb_private::Address ObjectFileELF::GetEntryPointAddress() {
 
 Address ObjectFileELF::GetBaseAddress() {
+  if (GetType() == ObjectFile::eTypeObjectFile) {
+    for (SectionHeaderCollIter I = std::next(m_section_headers.begin());
+         I != m_section_headers.end(); ++I) {
```
For my own understanding, why do we start at the 2nd section header?

clayborg replied:
> For my own understanding, why do we start at the 2nd section header?

The first one is all zeroes in every ELF file

bulbazord commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.h:
```diff
@@ -0,0 +1,165 @@
+#ifndef LLDB_SOURCE_PLUGINS_DYNAMICLOADER_FREEBSD_KERNEL_DYNAMICLOADERFREEBSDKERNEL_H
```
This file needs a license header.

bulbazord commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+ThreadPlanSP DynamicLoaderFreeBSDKernel::GetStepThroughTrampolinePlan(
+    lldb_private::Thread &thread, bool stop_others) {
+  Log *log = GetLog(LLDBLog::Step);
+  LLDB_LOGF(log, "Could not find symbol for step through.");
```
Is this log useful to you? Rather than saying "Couldn't find symbol for step through", it might be more accurate to say "DynamicLoaderFreeBSDKernel::GetStepThroughTrampolinePlan is not yet implemented."

If the log isn't useful though, I'd recommend removing it.

bulbazord commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+    }
+
+    // Set name for kernel
+    ConstString kernel_name("freebsd_kernel");
```
Please avoid using `ConstString` unless you're calling a method which requires it. It looks like this is just to set a `std::string` in the `KModImageInfo` below. I'd recommend using `llvm::StringRef` here.

For context, anything you put into the `ConstString` string pool persists until LLDB exits. There is no way to reclaim that memory when you're done with it. This is useful in some cases (e.g. you have the same string a ton of times and will use them throughout the debug session). In this case, `kernel_name` only sticks around for a short amount of time.

bulbazord commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+    if (m_linker_file_head_addr.IsValid()) {
+      if (!ParseKmods(m_linker_file_head_addr))
+        m_linker_files_list.clear();
+      return true;
```
What does this `return true` indicate? The `ParseKmods` invocation 2 lines above this could fail and all of the linker files could be blown away with the following call to `clear`. Is that not a condition you would want to consider?

aokblast replied:
Oh, it does nothing for now because I don't do anything when the kmod reading failed.
I leave it as a remind if I want to log anything when this function execute failed.
I don't know if I should modify it to return void or just reserve it.

bulbazord commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+    kld_load_addr =
+        m_process->ReadPointerFromMemory(current_kld + kld_off_address, error);
+
+    kmods_list.emplace_back();
+    KModImageInfo &kmod_info = kmods_list.back();
+    kmod_info.SetName(kld_filename);
+    kmod_info.SetLoadAddress(kld_load_addr);
+    kmod_info.SetPath(kld_pathname);
+    current_kld =
+        m_process->ReadPointerFromMemory(current_kld + kld_off_next, error);
```
Shouldn't you check the result of those memory reads above before mutating state here? Otherwise you may end up with stale or otherwise incorrect information, no?

clayborg replied:
> Shouldn't you check the result of those memory reads above before mutating state here? Otherwise you may end up with stale or otherwise incorrect information, no?

This will return LLDB_INVALID_ADDRESS if it fails, so you can check the error if you want, or just need to deal with LLDB_INVALID_ADDRESS correctly.

clayborg replied:
But yes, check if you got valid strings back for "kld_filename", "kld_pathname" and "kld_load_addr != LLDB_INVALID_ADDRESS would be good idea before adding the to the kmods_list. The "current_kld" should be checked against zero and LLDB_INVALID_ADDRESS in the while loop

aokblast replied:
Can I just check the error that was passed into the function?

bulbazord replied:
Either one is fine, you could also do both? This is one place where LLDB can be a little frustrating because there are 2 ways to detect the error and it's not obvious which one you should be checking.

bulbazord commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+  if (m_load_address == LLDB_INVALID_ADDRESS)
+    return false;
+
+  FileSpec file_spec(m_name.c_str());
```
You can create an `llvm::StringRef` out of a `std::string`. Creating one from a `const char *` involves doing a `strlen`.

emaste commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.h:
```diff
@@ -0,0 +1,165 @@
+  static void Terminate();
+
+  static llvm::StringRef GetPluginNameStatic() {
+    return "DynamicLoaderFreeBSDKernel";
```
It's always one word, FreeBSD, so "freebsd-kernel"

bulbazord commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+
+  // The kernel binary is from memory
+  if (this_is_kernel) {
+    if (log)
```
No need to check if the log object is valid, the macro does it for you already.

bulbazord commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+            addr);
+
+  llvm::ELF::Elf32_Ehdr header;
+  if (!ReadELFHeader(process, addr, header))
+    return UUID();
```
Should we not also set `read_error` here too? and some of the places below.

emaste commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+    return false;
+  if (objfile->GetType() != ObjectFile::eTypeExecutable)
+    return false;
+  if (objfile->GetStrata() != ObjectFile::eStrataUnknown &&
+      objfile->GetStrata() != ObjectFile::eStrataUser)
```
We can identify a FreeBSD kernel by the bogus PT_INTERP it specifies:
```
  INTERP         0x000238 0xffffffff80200238 0x0000000000200238 0x00000d 0x00000d R   0x1
      [Requesting program interpreter: /red/herring]
```

There are a bunch of ELF notes, but nothing that would uniquely identify it as a FreeBSD kernel. I suspect the Xen notes are unique to kernels, but won't be present if built without Xen support I guess. (As an aside I guess we should teach llvm-readelf about those note types.)

```
Displaying notes found in: .note.gnu.build-id
  Owner                Data size        Description
  GNU                  0x00000014       NT_GNU_BUILD_ID (unique build ID bitstring)
    Build ID: 984b7de7a87159b85761dc6bdc9645e28407e890

Displaying notes found in: .note.Xen
  Owner                Data size        Description
  Xen                  0x00000008       Unknown note type: (0x00000006)
   description data: 46 72 65 65 42 53 44 00
  Xen                  0x00000009       Unknown note type: (0x00000007)
   description data: 30 78 31 33 64 36 37 37 00
  Xen                  0x00000008       Unknown note type: (0x00000005)
   description data: 78 65 6e 2d 33 2e 30 00
  Xen                  0x00000008       Unknown note type: (0x00000003)
   description data: 00 00 00 80 ff ff ff ff
  Xen                  0x00000008       Unknown note type: (0x00000004)
   description data: 00 00 00 00 00 00 00 00
  Xen                  0x00000008       NT_VERSION (version)
   description data: 00 b0 03 81 ff ff ff ff
  Xen                  0x00000008       NT_ARCH (architecture)
   description data: 00 a0 03 81 ff ff ff ff
  Xen                  0x00000008       Unknown note type: (0x0000000c)
   description data: 00 00 00 00 00 80 ff ff
  Xen                  0x0000005e       Unknown note type: (0x0000000a)
   description data: 77 72 69 74 61 62 6c 65 5f 64 65 73 63 72 69 70 74 6f 72 5f 74 61 62 6c 65 73 7c 61 75 74 6f 5f 74 72 61 6e 73 6c 61 74 65 64 5f 70 68 79 73 6d 61 70 7c 73 75 70 65 72 76 69 73 6f 72 5f 6d 6f 64 65 5f 6b 65 72 6e 65 6c 7c 68 76 6d 5f 63 61 6c 6c 62 61 63 6b 5f 76 65 63 74 6f 72 00
  Xen                  0x00000004       Unknown note type: (0x00000009)
   description data: 79 65 73 00
  Xen                  0x00000008       Unknown note type: (0x0000000d)
   description data: 01 00 00 00 01 00 00 00
  Xen                  0x00000008       Unknown note type: (0x00000008)
   description data: 67 65 6e 65 72 69 63 00
  Xen                  0x00000004       Unknown note type: (0x0000000e)
   description data: 00 00 00 00
  Xen                  0x00000004       Unknown note type: (0x0000000b)
   description data: 79 65 73 00
  Xen                  0x00000004       Unknown note type: (0x00000012)
   description data: 30 b0 03 01
```

Oh, I guess a section named `set_sysinit_set` could be a decent indication.
```
There are 61 section headers, starting at offset 0x19a0c38:

Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            0000000000000000 000000 000000 00      0   0  0
  [ 1] .interp           PROGBITS        ffffffff802002a8 0002a8 00000d 00   A  0   0  1
  [ 2] .hash             HASH            ffffffff802002b8 0002b8 02dc10 04   A  4   0  4
...
  [49] set_sysinit_set   PROGBITS        ffffffff81746070 1546070 003878 00  WA  0   0  8
```

There are a bunch of `set_*_set` sections but the SYSINT linker set is always going to be present.

clayborg replied:
yes, any of the above suggested checks would be fine. Anything that can help us return "eStrataKernel" in ObjectFileELF could be really nice to add.

emaste replied:
I discussed with our kernel gdb maintainer and there PT_INTERP == /red/herring is used to identify a kernel for the main executable.

clayborg replied:
> I discussed with our kernel gdb maintainer and there PT_INTERP == /red/herring is used to identify a kernel for the main executable.

That is great. Can we have ObjectFileELF.cpp check for this and return eStrataKernel for the ObjectFileELF::GetStrata() and then use this value to return true to the is_kernel() function?

aokblast replied:
Sorry about my late reply, I am preparing some programming contest these day. 
I don't know if we should consider other kernel like Linux in here. I will test it on Linux and make both kernel pass this test in ObjectFileELF.cpp.

clayborg replied:
It is ok if something only works for FreeBSD too, just check the OSABI in the ELF header if this only works for FreeBSD. If this works for linux as well, then feel free to make it work for both!

augusto2112 commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+  return "The Dynamic Loader Plugin For FreeBSD Kernel";
+}
+
+static bool is_kernel(Module *module) {
```
Nit: LLDB style would be `IsKernel`

aokblast replied:
Oh, I check the coding style in DarwinKernel DynamicLoader and found that it uses snake case in static function.

clayborg commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+  // relocatable file So what we do is to set the load address only.
+  if (is_kmod(m_module_sp.get()) && is_reloc(m_module_sp.get())) {
+    m_stop_id = process->GetStopID();
+    bool changed;
```
init this to false

clayborg commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:
```diff
@@ -0,0 +1,770 @@
+  addr_t current_kld =
+      linker_files_head_addr.GetLoadAddress(&m_process->GetTarget());
+
+  while (current_kld != 0) {
```
You will want to check "current_kld != LLDB_INVALID_ADDRESS && current_kld != 0". Both linker_files_head_addr.GetLoadAddress(...) and m_process->ReadPointerFromMemory(...) will return LLDB_INVALID_ADDRESS which is not zero, but you probably might read a NULL pointer to terminate the module list.

aokblast commented:
I fix all mentioned issue and test on both x86-64 and arm64.
Hope everyone can give it a review.
Thanks!

clayborg reviewed (commented):
Just one header file header to be fixed and this looks good to me. I will let other FreeBSD folks do the final approval since I don't work on FreeBSD.

aokblast commented on lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.h:
```diff
@@ -0,0 +1,172 @@
+//===-- DynamicLoaderFreeBSDKernel.h
+//------------------------------------------===//
```
Fix it

emaste commented:
Needs to be adapted for f2d32ddcec82c20582c6aa32558b82ca7c3d3c50, `lldb/source/{Core => Host/common}/StreamFile.cpp`

(I've applied the change locally for testing, and will just fold it into the patch series if otherwise ready to go.)

aokblast commented:
Excuse me. If the code works well now, I want to ask if I need to modify the commit message by myself so the message looks tidy, or let somebody landing this pull request modify the message?

aokblast commented:
Thanks, maybe aokblast <aokblast@FreeBSD.org> is more suitable. XD

emaste commented:
> I will let other FreeBSD folks do the final approval since I don't work on FreeBSD.

@clayborg this change LGTM for FreeBSD. I'm still not fully sorted on the approach for landing the commits post transition to GitHub. Also see the comment about author name above.

clayborg commented:
> > I will let other FreeBSD folks do the final approval since I don't work on FreeBSD.
> 
> @clayborg this change LGTM for FreeBSD. I'm still not fully sorted on the approach for landing the commits post transition to GitHub. Also see the comment about author name above.

There is a web page that covers the new submission process: https://llvm.org/docs/GitHub.html

I haven't done it yet either

emaste commented:
https://llvm.org/docs/GitHub.html#landing-your-change

> There are two different ways to do this:
>
> - [Interactive rebase](https://git-scm.com/docs/git-rebase#_interactive_mode) with fixup’s. This is the recommended method since you can control the final commit message and inspect that the final commit looks as you expect. When your local state is correct, remember to force-push to your branch and press the merge button afterwards.
>
> - Use the button Squash and merge in GitHub’s web interface, if you do this remember to review the commit message when prompted.

So I suspect the best way to go here is @aokblast locally squashes these commits into one, updates author if desired, fixes up commit messages if necessary, and then force-pushes to this pull request. Then @clayborg or I push the merge button.

emaste commented:
Oh, I think we should also mention GSoC in the commit message. Something like `This project was part of FreeBSD's participation in Google Summer of Code 2023.` or so.

aokblast commented:
As my mentor Li-Wen Hsu suggest, It is better to separate this patch into two commits. One for DynamicLoaderFreeBSDKernel class, and another for the modification about ObjectFile.cpp. Should I use 1 commits to include both of these change or separate it as two?

clayborg commented:
> As my mentor Li-Wen Hsu suggest, It is better to separate this patch into two commits. One for DynamicLoaderFreeBSDKernel class, and another for the modification about ObjectFile.cpp. Should I use 1 commits to include both of these change or separate it as two?

I am ok with it being one patch since it is related to implementing kernel debugging.

aokblast commented:
OK, I have modified the commit message.

emaste commented:
Hmm, when I attempt to close it via the GitHub UI I get:
![image](https://github.com/llvm/llvm-project/assets/1034582/54534c5d-9873-40c4-81af-45c50919dc92)

I think I can just fetch the commit locally and then push it to the tree

aokblast committed:
Add DynamicLoader Plugin Fore FreeBSD Kernel coredump
This patch add dynamicloader plguin for freebsd kernel coredump on lldb.
The implementation is by parsing linker_files structure and get all loaded kernel modules.

This patch was part of FreeBSD's participation in Google Summer of Code 2023

aokblast commented:
I try to change the committer of my patch and I think it works now.

DavidSpickett commented:
FYI on our Windows on Arm (aka AArch64) bot we got this warning:
```
[4840/6117] Building CXX object tools\lldb\source\Plugins\DynamicLoader\FreeBSD-Kernel\CMakeFiles\lldbPluginDynamicLoaderFreeBSDKernel.dir\DynamicLoaderFreeBSDKernel.cpp.obj
C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\source\Plugins\DynamicLoader\FreeBSD-Kernel\DynamicLoaderFreeBSDKernel.cpp(540,7): warning: format specifies type 'unsigned long' but the argument has type 'std::vector<DynamicLoaderFreeBSDKernel::KModImageInfo>::size_type' (aka 'unsigned long long') [-Wformat]
      linker_files_list.size());
      ^~~~~~~~~~~~~~~~~~~~~~~~
C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\include\lldb/Utility/Log.h(353,48): note: expanded from macro 'LLDB_LOGF'
      log_private->Formatf(__FILE__, __func__, __VA_ARGS__);                   \
                                               ^~~~~~~~~~~
1 warning generated.
```
And this one on the 32 bit Arm Linux bot:
```
../llvm-project/lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp:540:7: warning: format specifies type 'unsigned long' but the argument has type 'size_type' (aka 'unsigned int') [-Wformat]
1 warning generated.
```
AArch64 Linux was fine. So I assume each one is using one of `unsigned int`, `unsigned long` or `unsigned long long` for its size types.

There is probably a way to print that in a portable way but I don't recall it at the moment.

aokblast commented:
I think the answer is %z from cpp reference as the following screenshot shows.
I will check the whole code to prevent issues like this.

![CleanShot 2023-10-04 at 19 17 55](https://github.com/llvm/llvm-project/assets/16476727/8ffdc723-18c5-410a-84d9-40bf0620dd6f)

emaste commented:
I'm not sure why @aokblast's reply didn't appear here, but indeed `%zu` is the right format specifier.

aokblast commented:
I send another PR in [68210](https://github.com/llvm/llvm-project/pull/68210). I don't know if it is the correct way to fix this problem.
